### ISSUE TITLE
Get Servo working on Windows

### ISSUE THREAD

**Issue by kmcallister** [servo/servo#1908]
Stuff we need to implement for sure:

* Build system (Rust uses MinGW/MSYS; we can do the same)
* Font loading
* Sending OpenGL textures between GL contexts and (eventually) between processes

And then there's probably a lot of other work we don't know about.

See `src/platform/` and `src/support/layers/rust-layers/platform/`.

Rust [already works](https://github.com/mozilla/rust/wiki/Using-Rust-on-Windows) but some of the things we use may be missing.

We will never, ever support Windows XP.

**Comment by Manishearth** [comment]
Going to use this as a tracking issue

**Comment by Manishearth** [comment]
Requirements:

Native bindings:

**Comment by Manishearth** [comment]
cc @klutzy @retep998

**Comment by Manishearth** [comment]
Makefile notes  (https://github.com/servo/rust-harfbuzz/pull/36, https://github.com/servo/mozjs/pull/27): 

Check for Windows (`ifeq (windows,$(findstring windows,$(TARGET)))`), then do the following:

 - `OUT_DIR:=$(shell cygpath "$(OUT_DIR)")`
 -  `ifeq (cc,$(CC))`, set `CC = gcc`, `CPP = gcc -E`

**Comment by retep998** [comment]
On Windows we can abandon OpenSSL and use WinHTTP which is a native system library. It is what Cargo uses on Windows. But since MSYS2 can provide us with an OpenSSL package, it isn't too much of a hassle.
https://msdn.microsoft.com/en-us/library/windows/desktop/aa382925%28v=vs.85%29.aspx
As for font loading, use the native Windows methods?
https://msdn.microsoft.com/en-us/library/dd162620%28v=vs.85%29.aspx

**Comment by Manishearth** [comment]
We'd need to make Hyper and Servo SSL-backend-agnostic then, which sounds like more work.

Besides, the handling of cookies depends on libcrypto. If we get openssl on msys2 that should be enough, otherwise it's not too hard to just build it (we do this for Android already)

Native windows methods sound good. Even better if we can wrap it in an API that's like fontconfig or something.

**Comment by tomaka** [comment]
> [ ] Glutin (should just work)

The windows implementation of glutin is actually the one that I trust the most.

**Comment by cybergeek94** [comment]
I would like to volunteer [fontconfig-rs][1] for cross-platform font loading. It doesn't *technically* have OS X or Windows support yet but it's really just a matter of having it pass a custom config to the `fontconfig` library that tells it where the fonts live on those platforms. GIMP is a good example of a program using fontconfig on Windows.

[1]: https://github.com/cybergeek94/fontconfig-rs

**Comment by retep998** [comment]
@cybergeek94 No but I can easily add any methods needed.

**Comment by vvuk** [comment]
For fontconfig, is what's needed just font selection, which is then passed to Skia for shaping and rasterization?  Win32 fontconfig already exists as part of the win32 gtk+ port as was mentioned above, so it may be possible to just use that along with fontconfig-rs.  Bundling all these DLLs (openssl, fontconfig, etc.) together in a "servo-win32-support" package would probably be helpful so that people don't have to download everything individually.

I can help with the GL texture sharing cross context/process stuff.

**Comment by metajack** [comment]
Won't not using platform font enumeration methods result in users seeing inconsistent fonts between native platform apps and Servo? Or is fontconfig using native facilities behind the scenes?

**Comment by vvuk** [comment]
It will; I would expect that along with stuff like openssl, it would be a stopgap measure in order to get a port up.  Then pieces can be rewritten to use the native Windows APIs,or at least to work in a more Windows-like manner. On Windows, Gecko uses DirectWrite natively, though font selection is mostly done by Gecko directly and less by fc.  (Fc gets used for enumeration/querying/etc.)

**Comment by cybergeek94** [comment]
@metajack Windows stores all its fonts in pretty much one location (`C:\Windows\Fonts`), so there shouldn't be a discrepancy between the selection queried through the Win API and fontconfig. The Unix-based platforms actually have a bit more fragmentation in this area; OS X is really the only complex platform for font location since on Linux, fontconfig is installed through the given package manager for a distro with a configuration file that tells it where the fonts live in that particular distro.

**Comment by retep998** [comment]
Not all fonts on Windows are actually stored in `C:\Windows\Fonts`. Fonts can be added and removed at any time using `AddFontResource`, where they will be listed in the fonts when you use the Windows API for enumerating fonts, but they won't actually be in the Fonts folder. A `WM_FONTCHANGE` will be sent to all top level windows to indicate of any such additions/removals as well.

**Comment by cybergeek94** [comment]
@retep998 I wouldn't depend on those fonts being available for any respectable amount of time if they can be added and removed at will. I don't think Servo should, either. FWICT, installed custom fonts are copied into this folder so they are available at any time, which would be my only concern.

**Comment by retep998** [comment]
Fonts can be added or removed from the Fonts folder at will as well, and we'd need to hook into the notifications regardless to detect that. If we want to be robust, we have to take the approach Firefox uses. If we don't want to be robust for now, just to get things working, that's fine. But eventually this is something that we'll need to do.

**Comment by Manishearth** [comment]
What does Firefox do here?

**Comment by Manishearth** [comment]
```
07:40 < nattokirai> gecko builds lists of font families on the system
07:40 < nattokirai> then fills in the details as needed
07:41 < Manishearth> nattokirai: any library we can use?
07:41 < nattokirai> for windows there are two backends, one using DirectWrite, the other GDI API's
...
07:42 < nattokirai> um, "library" meaning modular piece of code that can simply be imported into servo? no... :P
07:42 < Manishearth> dammit
07:42 < nattokirai> but...
07:42 < nattokirai> the directwrite API is fairly sane
...
07:43 < nattokirai> look through the code in gfx/thebes
```

I guess we should have a look at DirectWrite?

**Comment by nattokirai** [comment]
Not sure what servo is doing for CSS font matching but that code really needs to exist in servo, since you need an implementation for on-demand loading for @font-face families. You **really** want to use native platform API's and **not** fontconfig for this. So DirectWrite on Windows, CoreText on OSX, fontconfig on Linux and roll-your-own on Android.

**Comment by cybergeek94** [comment]
Sounds like a new library project: cross-platform font querying using native APIs.

**Comment by nattokirai** [comment]
Yeah, the level of support for various things related to fonts needed by a modern web browser varies across platforms. In general, you'll need to maintain some sort of font lookup interface and fill in what's needed on a per-platform basis. OSX and DirectWrite are fairly easy, fontconfig and Android are relatively painful. Gecko uses platform API's to maintain a list of font families available and fills in data for a given family when it's actually used. Font style matching really needs to be done by servo code and not punted to platform-level API's which have their own rules for style matching.

**Comment by nattokirai** [comment]
Key features of a font system interface:

- lookup a font **family** name
- enumerate styles within a single family
- select a given face within a family based on CSS font properties (i.e. weight/width/slant)
- activate a font from data (with required sanitation such as OTS)
- lookup individual font faces via a fullname or Postscript name

**Comment by nattokirai** [comment]
For browsers you'll also need:

- system font fallback, the ability to find a font for a given character
- mappings of generics based on language/script (e.g. ``sans-serif`` for Arabic and Japanese are not the same font)
- ability to read (and cache) font metrics and font table data

**Comment by Manishearth** [comment]
cc @SimonSapin

**Comment by kmcallister** [comment]
I've created a new ticket for the font discussion, which is drifting away from Windows-specific.

**Comment by vvuk** [comment]
From poking at this, a large chunk of the difficulty in getting Servo building on Windows will be dealing with paths.  Specifically, os::Path on windows normalizes all paths to true Windows paths, e.g. "C:\foo\bar".  This is a problem, because a lot of existing crate makefiles will do things like "cd $(OUTDIR) && ...", which will turn into "cd c:\foo\bar", which becomes "cd c:foobar" when it gets executed.

I don't have any good suggestions on how to fix this; running through and fixing all the makefiles and other build system bits seems like a losing proposition.  The easiest thing that I can think of is to introduce yet another path type, msys-type paths, and then force use of them when running on Windows under MSYS.  These would be "c:/foo/bar" (not /c/foo/bar -- c:/foo/bar can often be passed to native tools like visual studio stuff and it'll work just fine, whereas /c/foo/bar will just confuse it).  

The next thing would be to go through and replace \ with / in any paths that get turned into strings within cargo and cargo-built build scripts.  Not sure if that can be forced somehow at compile time.

Regardless, it seems that fixing this will need to be done at some low level where MSYS is detected.

**Comment by Manishearth** [comment]
We could just do all path manipulation in the build script and pass these full paths as env vars to the build script. Path manipulation can be done in an os independent way in Rust, afaict.

-----Original Message-----

**Comment by SimonSapin** [comment]
> a lot of existing crate makefiles will do things like "cd $(OUTDIR) && ...", which will turn into "cd c:\foo\bar", which becomes "cd c:foobar" when it gets executed.

Isn’t this fixed with quoting? `cd "$(OUTDIR)" && ...`

**Comment by vvuk** [comment]
> Isn’t this fixed with quoting? cd "$(OUTDIR)" && ...

It is, but requires all the relevant crate build scripts to quote appropriately, or things will break on Windows.  If there was a unified monolithic build system (e.g. how Gecko is built) then that seems reasonable, but given the (better) modular approach, I don't think there's a way of enforcing Windows build compatibility.

**Comment by klutzy** [comment]
@vvuk: `OUT_DIR:=$(shell cygpath "$(OUT_DIR)")` is used in servo/rust-harfbuzz#36 and  servo/mozjs#27. Cargo sets `OUT_DIR` as environment variable, so the simple trick works.
This is msys-specific issue and I think Rust or Cargo shouldn't need to understand msys.

**Comment by vvuk** [comment]
> @vvuk: OUT_DIR:=$(shell cygpath "$(OUT_DIR)") is used in servo/rust-harfbuzz#36 and servo/mozjs#27. Cargo sets OUT_DIR as environment variable, so the simple trick works. This is msys-specific issue and I think Rust or Cargo shouldn't need to understand msys.

Yep, that works fine -- but it requires every single makefile.cargo to check for windows and include `OUT_DIR:=$(shell cygpath "$(OUT_DIR)")` or it will just simply fail to work on Windows.  That seems counter to making things easy. MSYS is a unique environment beyond just plain Windows; treating it as one can avoid a lot of future cargo pain in general, not just Servo.

Another option is to do away with makefiles entirely and use a different build system that's more robust.  But I don't have any good suggestions -- a pure rust-based system would probably be best, but I don't know if such a thing exists.

I guess a third option may be to have cargo prepend a "windows preamble" to every makefile.cargo, or make that into a generic feature -- maybe the ability to prepend a preamble to makefile.cargo, or even to have something act as a runner for build.rs build tools that can munge the env var in between what cargo sets up and what the build script sees.

**Comment by vvuk** [comment]
> We could just do all path manipulation in the build script and pass these full paths as env vars to the build script. Path manipulation can be done in an os independent way in Rust, afaict.

Rust seems to support windows "C:\foo\bar" paths and Unix "/foo/bar" paths.  I don't see any support for the desired hybrid "C:/foo/bar" syntax that msys and a lot of native tools understand.

**Comment by jasonwilliams200OK** [comment]
> I don't think Windows has this library natively. Either we build it, or we download ..

Maybe you can seek some pointers here https://github.com/microsoft/openssl. @khouzam, can you provide some help on the easiest path to build OpenSSL lib for classic and old desktops?

**Comment by jasonwilliams200OK** [comment]
BTW, you can try building with VS2013 community version (for academics and open-source projects): https://www.visualstudio.com/en-us/products/visual-studio-community-vs. Rust support is provided by Mozilla herself via their official Nuget package: https://www.nuget.org/packages/Rust/.

**Comment by khouzam** [comment]
Thanks for reaching out.

The instructions to build OpenSSL are pretty straightforward. You can look at the latest on https://github.com/openssl/openssl/blob/master/INSTALL.W32 

It does require Perl and supports Visual C++, Borland and GCC.

In 3 steps for Visual Studio it would be:
1. git clone https://github.com/openssl/openssl.git
2. perl Configure VC-WIN32 no-asm --prefix=c:/some/openssl/dir
3. ms\do_ms

From a VC++ prompt
4. nmake -f ms\ntdll.mak 

And you should have the DLL version of OpenSSL built and available.

**Comment by jasonwilliams200OK** [comment]
That's it! Thank you! :+1:

**Comment by Manishearth** [comment]
I fixed up the png compile.

https://github.com/servo/rust-png/pull/73

**Comment by Manishearth** [comment]
A relatively simple task that can be done here is making `mach` `cygpath` friendly when it sets the PATH environment variable amongst other things.

**Comment by Manishearth** [comment]
Fixed up stb too

https://github.com/servo/rust-stb-image/pull/71

**Comment by Manishearth** [comment]
Skia seems to need MSVC to build on windows

There are some instructions for getting it to work on mingw, but they seem dated:

https://groups.google.com/forum/#!topic/skia-discuss/LO1kdT0tmNs

https://code.google.com/p/skia/issues/detail?id=213

**Comment by waddlesplash** [comment]
Does Rust use a new enough LLVM that the MSVC ABI can be used? That would solve the problem without requiring MinGW.

**Comment by Manishearth** [comment]
https://github.com/rust-lang/rust/pull/25350

Rust can build on MSVC, it's just new. And while most things will just work with mingw, the same is not true for msvc.

**Comment by Manishearth** [comment]
OpenSSL seemed to magically build for me on mingw64. I'm ticking it off, but it's possible that this was just my setup.

**Comment by indivisible** [comment]
I've tried to cross-compile servo on a Linux host to Windows. What I did:
- set up a simple mingw environment (I used https://github.com/lachs0r/mingw-w64-cmake)
- clone servo
- clone the rust version it needs, build it with windows support, and put it somewhere the build process will find
- try to build servo

I got a bunch of bugs with some cargo packages, which I've reported and worked around myself.
For some of the problem packages (fontconfig, openssl) I just hacked the build to link the glue code to libraries I built with the mingw env's packages.
The major problem I got stuck at is that the Skia changes for Servo are NOT ported to Windows yet (only OS X, Linux and Android are supported.)

The cross-compile process is mildly painful, but not impossible. (Although I have some pretty strong words for whoever keeps inventing new build systems at Mozilla.)

**Comment by vvuk** [comment]
Can you shove your hacks and changes in a gist?  I'd like to work from there.  I can definitely port the Skia changes over etc., but need to get through the random cargo etc. breakage with fontconfig/openssl/etc.

**Comment by vvuk** [comment]
Note that I think that it's acceptable to have an initial Servo-on-Windows have dependencies such as: requires msys2/mingw64 build env; requires clang-cl (or real msvc) to build some modules; etc.  We can then remove them over time but at least we have something that can be hacked on.

**Comment by pcwalton** [comment]
Note that we'll need some kind of `ipc-channel` port now.

**Comment by vvuk** [comment]
Writing the code is the super easy part of all this.  I will happily write code all day long.  It's fighting with build system cruft that's the painful bit..

**Comment by waddlesplash** [comment]
> Note that we'll need some kind of ipc-channel port now.

Windows' named pipes? They aren't very secure (if you know the name of the pipe, you can read stuff off of it, essentially) but they're tried, true, and fast.

**Comment by Jayflux** [comment]
is there any low-hanging-fruit i can help with?

**Comment by Manishearth** [comment]
Not that I know of, though @vvuk might have some tasks he can delegate

**Comment by retep998** [comment]
Someone can always go write a Rust library around DirectWrite, especially since there are raw bindings to it.

**Comment by vvuk** [comment]
My current plan is to get things compiling, stubbing out as necessary along
the way even if the final result doesn't work. Then it will be much easier
to parallelize getting things to work, since people won't be fighting the
build the entire time. I hope to have things compiling in the next day or
two, then a day of submitting PRs and replicating the work from scratch.

**Comment by waddlesplash** [comment]
> Someone can always go write a Rust library around DirectWrite, especially since there are raw bindings to it.

FWIW, I *really* do not like the way Chrome (at least) looks with DirectWrite turned on. Maybe that's Chrome's fault and not DirectWrite, but I prefer Firefox's font rendering on Windows by a lot. Ubuntu's default font rendering settings surpass it by a lot, though.

**Comment by retep998** [comment]
>Maybe that's Chrome's fault and not DirectWrite, but I prefer Firefox's font rendering on Windows by a lot. 

Firefox also uses DirectWrite on Windows. There's a ton of tweaking you can do to adjust how cleartype renders text.

**Comment by waddlesplash** [comment]
#7878.

**Comment by larsbergstrom** [comment]
Landed in #9385!
